rust-lang/cargo
# fix: encode URL params correctly for SourceId in Cargo.lock

### What does this PR try to resolve?

Fixes #11085, also as a part of #12120.


When serialize `Resolve` for `Cargo.lock`, previously Cargo relys on `SourceId::as_url`, which is good for user-facing string. However, it leads to an incorrect deserialization since `url` crate always assume the given string is correctly url-encoded.

https://github.com/rust-lang/cargo/blob/0d5370a9ae0e97f000e63e416bafbf2abb57e41d/src/cargo/core/source/source_id.rs#L527-L538

This fixes by adding a wrapper for encoding `SourceId` during the serialization for `Cargo.lock`.


### How should we test and review this PR?

This is a conitnuation of #12279. #12279 should be reviewed first.


We use [`form_urlencoded::byte_serialize`](https://docs.rs/form_urlencoded/1.2.0/form_urlencoded/fn.byte_serialize.html), which is re-exported by `url` crate. Tests are copied from <https://github.com/rust-lang/cargo/pull/11086>. Kudos to the original author!
